drivers/sensor/st/lsm9ds1/lsm9ds1.c

```diff
@@ -0,0 +1,651 @@
+
+/* Sensitivity of the accelerometer, indexed by the raw full scal value. Unit is µg/ LSB */
+static const uint16_t lsm9ds1_accel_fs_sens[] = {61, 732, 122, 244};
+
+/*
+ * Sensitivity of the gyroscope, indexed by the raw full scale value.
+ * The value here is just a factor applied to GAIN_UNIT_G, as the sensitivity is
+ * proportional to the full scale size.
+ * The index 2 is never used : the value `0` is just a placeholder.
+ */
+static const uint16_t lsm9ds1_gyro_fs_sens[] = {1, 2, 0, 8};
+
+/*
+ * Values of the different sampling frequencies of the accelerometer, indexed by the raw odr
+ * value that the sensor understand.
+ */
+static const uint16_t lsm9ds1_odr_map[] = {0, 10, 50, 119, 238, 476, 952};
+
+/*
+ * Value of the different sampling frequencies of the gyroscope, indexed by the raw odr value
+ * that the sensor understand.
+ */
+static const uint16_t lsm9ds1_gyro_odr_map[] = {0, 15, 59, 119, 238, 476, 952};
+
+static int lsm9ds1_reboot(const struct device *dev)
+{
+	const struct lsm9ds1_config *cfg = dev->config;
+	stmdev_ctx_t *ctx = (stmdev_ctx_t *)&cfg->ctx;
+	lsm9ds1_ctrl_reg8_t ctrl8_reg;
+	int ret;
+
+	ret = lsm9ds1_read_reg(ctx, LSM9DS1_CTRL_REG8, (uint8_t *)&ctrl8_reg, 1);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ctrl8_reg.boot = 1;
+
+	ret = lsm9ds1_write_reg(ctx, LSM9DS1_CTRL_REG8, (uint8_t *)&ctrl8_reg, 1);
+	if (ret < 0) {
+		return ret;
+	}
+
+	k_msleep(50);
+
+	return 0;
+}
+
+static int lsm9ds1_accel_range_to_fs_val(int32_t range)
+{
+	switch (range) {
+	case 2:
+		return LSM9DS1_2g;
+	case 4:
+		return LSM9DS1_4g;
+	case 8:
+		return LSM9DS1_8g;
+	case 16:
+		return LSM9DS1_16g;
+	default:
+		return -EINVAL;
+	}
+}
+
+static int lsm9ds1_gyro_range_to_fs_val(int32_t range)
+{
+	switch (range) {
+	case 245:
+		return LSM9DS1_245dps;
+	case 500:
+		return LSM9DS1_500dps;
+	case 2000:
+		return LSM9DS1_2000dps;
+	default:
+		return -EINVAL;
+	}
+}
+
+static int lsm9ds1_accel_fs_val_to_gain(int fs)
+{
+	return lsm9ds1_accel_fs_sens[fs];
+}
+
+static int lsm9ds1_accel_freq_to_odr_val(uint16_t freq)
+{
+	size_t i;
+
+	for (i = 0; i < ARRAY_SIZE(lsm9ds1_odr_map); i++) {
+		if (freq <= lsm9ds1_odr_map[i]) {
+			return i;
+		}
+	}
+
+	return -EINVAL;
+}
+
+static int lsm9ds1_gyro_freq_to_odr_val(uint16_t freq)
+{
+	size_t i;
+
+	for (i = 0; i < ARRAY_SIZE(lsm9ds1_gyro_odr_map); i++) {
+		if (freq <= lsm9ds1_gyro_odr_map[i]) {
+			return i;
+		}
+	}
+
+	return -EINVAL;
+}
+
+static int lsm9ds1_accel_set_odr_raw(const struct device *dev, uint8_t odr)
+{
+	const struct lsm9ds1_config *cfg = dev->config;
+	stmdev_ctx_t *ctx = (stmdev_ctx_t *)&cfg->ctx;
+	struct lsm9ds1_data *data = dev->data;
+	int ret;
+
+	lsm9ds1_ctrl_reg6_xl_t ctrl_reg6_xl;
+
+	ret = lsm9ds1_read_reg(ctx, LSM9DS1_CTRL_REG6_XL, (uint8_t *)&ctrl_reg6_xl, 1);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ctrl_reg6_xl.odr_xl = odr;
+
+	ret = lsm9ds1_write_reg(ctx, LSM9DS1_CTRL_REG6_XL, (uint8_t *)&ctrl_reg6_xl, 1);
+	if (ret < 0) {
+		return ret;
+	}
```

Comment:
I didn't use lsm9ds1_imu_data_rate_set, because I would have to pack the data_rate (for the accel and gyro) into the enum lsm9ds1_imu_odr_t, just for the function to unpack it after, so I figured it was better to avoid useless work (the shifts and mask, but also the useless register access in case we just want to change the accel).

But I use the function at the initialization, because the value in the device-tree property imu-odr is already in the correct format.